### PR TITLE
Keep value of controlled number fields upon action

### PR DESCRIPTION
**Pull request by vordgi**
## Summary

The second attempt to fix the problem. The first one (#30737) is simpler and just an addition of a workaround. Now I've approached it more comprehensively.

Improved the system for inserting default values. Previously, they were inserted sometimes on change, sometimes on blur, _sometimes never_. I've implemented logic for determining default values for number and email inputs before submit (_it's strange that there was a comment about the email input problem, but no condition_).

I've expanded checks for other input types (_it's strange that, for example, a `defaultValue` could theoretically be set for **radio**_)

In doing so, I removed the test "`should properly control 0.0 for a number input`". _I don't understand why it even exists - why is it expected that a controlled field can be overwritten in this way? It feels like it was working by sheer miracle_. But we can bring back the `ChangeEventPlugin` logic. However, as I've come to understand, it's also full of problems and seems quite prone to bugs. Plus, it's unnecessary actions

## Test plan

- added tests
- https://codesandbox.io/p/sandbox/react-keep-value-of-controlled-number-fields-upon-action-3sd2ly?file=%2Fsrc%2Findex.js

Closes https://github.com/facebook/react/issues/29862

### CODE REVIEW

**Comment by eps1lon** [comment]
Seems like this is failing with `yarn test --variant=true`

**Comment by vordgi** [comment]
While making the demo, I decided to check how checkable inputs work. And as it turned out, they don't work. I'll try to deal with them separately later. It's strange that there are no bugs reported for this... For clarity, of course, I checked in the version without my changes - https://codesandbox.io/p/devbox/silly-chebyshev-rshrhl-forked-g68h27.

And here's the version with my changes - https://codesandbox.io/p/devbox/656qfj. The numeric field works well, but the checkable fields are still being cleared:

https://github.com/user-attachments/assets/82c1c48e-1869-4d59-8852-cdf14af4b955

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js [hunk 1]
```diff
@@ -1002,29 +1002,6 @@ describe('ReactDOMInput', () => {
     expect(node.value).toBe('0');
   });
 
-  it('should properly control 0.0 for a number input', async () => {
```
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMInput-test.js:1005]: Why remove these tests? The new tests is tied to Form Actions so it would make sense to me to keep these tests.

File: packages/react-dom-bindings/src/client/ReactDOMInput.js [hunk 2]
```diff
@@ -418,26 +417,28 @@ export function restoreControlledInputState(element: Element, props: Object) {
...
-// https://github.com/facebook/react/issues/7253
+// Here we check to see if the defaultValue has actually changed and update it
 export function setDefaultValue(
   node: HTMLInputElement,
   type: ?string,
-  value: ToStringValue,
+  value: mixed,
+  force: ?boolean,
 ) {
+  if (['button', 'file', 'hidden', 'image', 'reset', 'submit'].includes(type))
```
Comment (eps1lon) [inline packages/react-dom-bindings/src/client/ReactDOMInput.js:427]: We prefer simple strict equality checks and `{}` around for our if blocks. It's more familiar to us.

File: packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js [hunk 3]
```diff
@@ -92,6 +96,20 @@ function extractEvents(
   }
   const formInst = maybeTargetInst;
   const form: HTMLFormElement = (nativeEventTarget: any);
+
+  if (!disableInputAttributeSyncing) {
+    // eslint-disable-next-line no-for-of-loops/no-for-of-loops
+    for (const element of form.elements) {
```
Comment (eps1lon) [inline packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js:102]: Do we iterate through DOM collections with for-of in other places as well?

**Comment by eps1lon** [comment]
FYI you can just use URLs in the package.json in the sandboxes. This helps to confirm the correct version is used.

Can you summarize in the PR description what is being fixed and attach a before/after video? I can't seem to spot a difference but maybe I'm using the wrong browser?

File: packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js (hunk 3 above)
Comment (vordgi) [reply packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js:102]: 686 times the for...of construction in repo (_36 with disable-next-line_), in some places specifically for DOMCollection:
* https://github.com/facebook/react/blob/3e94bce765d355d74f6a60feb4addb6d196e3482/packages/react-devtools-extensions/src/panel.js#L15
* https://github.com/facebook/react/blob/8fbd3079425eaff85ec1b52b0eefecdd44ba7289/packages/react-devtools-extensions/src/main/cloneStyleTags.js#L5

But this is devtools, I haven't encountered such a case specifically for react-dom

**Comment by vordgi** [comment]
> Can you summarize in the PR description what is being fixed and attach a before/after video? I can't seem to spot a difference but maybe I'm using the wrong browser?

This PR is specifically for fixing the number input. When submitting from the number input itself (_i.e., changing the value without leaving it and pressing Enter_) - the value won't reset to the initial one. In the previous version, it was resetting because it was saved on blur.

https://github.com/user-attachments/assets/f17a8e3f-26f8-4aee-9b86-da6de6a05a70

Now it is saved before submit:


https://github.com/user-attachments/assets/a6eb0153-1908-4341-aea7-610f843aecf3

**Comment by eps1lon** [comment]
Gotcha, thank you for the clarification. I was submitting via click and then it didn't reset. That's also probably why we haven't gotten any reports yet because most devs test with a mouse or with an explicit submit. Not by an implicit submit via {enter} while focusing a form element.

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 1 above)
Comment (vordgi) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js:1005]: Yes, I thought for a long about what to do with this test and left it as text in the PR description:

> In doing so, I removed the test "`should properly control 0.0 for a number input`". I don't understand why it even exists - why is it expected that a controlled field can be overwritten in this way? It feels like it was working by sheer miracle. But we can bring back the `ChangeEventPlugin` logic. However, as I've come to understand, it's also full of problems and seems quite prone to bugs. Plus, it's unnecessary actions

It turns out this test is no longer relevant, as the value doesn't change on blur anymore. Moreover, I don't understand why it should have changed in a controlled field in this way? It doesn't change like this anywhere else...

File: packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js (hunk 3 above)
Comment (eps1lon) [reply packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js:102]: devtools has less restrictions. Do we iterate through DOM collections somewhere else in react-dom?

**Comment by vordgi** [comment]
> Gotcha, thank you for the clarification. I was submitting via click and then it didn't reset. That's also probably why we haven't gotten any reports yet because most devs test with a mouse or with an explicit submit. Not by an implicit submit via {enter} while focusing a form element.

Yes, but at the same time this opens up some very unpleasant artifacts. And as I wrote above, I also caught the same error with checkable inputs (_which will try to fix a bit later_)

_Video without current fix:_

https://github.com/user-attachments/assets/fbdaf95e-323e-467a-8d6d-3bf5ba87feb0

File: packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js (hunk 3 above)
Comment (vordgi) [reply packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js:102]: Nope, "_But this is devtools, I haven't encountered such a case specifically for react-dom_"

Comment (eps1lon) [reply packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js:102]: We don't use `for-of` in react-dom-bindings so let's just stick with a classic `for (let i ...)` loop for safety.

Comment (vordgi) [reply packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js]: ✅

File: packages/react-dom-bindings/src/client/ReactDOMInput.js (hunk 2 above)
Comment (vordgi) [reply packages/react-dom-bindings/src/client/ReactDOMInput.js]: ✅

**Comment by vordgi** [comment]
During testing noticed existing bugs (_the same, but with checkbox and radio inputs_)

File: packages/react-dom/src/__tests__/ReactDOMForm-test.js [hunk 4]
```diff
@@ -1590,6 +1592,95 @@ describe('ReactDOMForm', () => {
...
+    expect(numberInputRef.current.value).toBe('2024');
+
+    await act(() => resolveText('Done'));
+    assertLog(['Async action ended']);
+
+    if (disableInputAttributeSyncing) {
+      // The DOM resets after action end when "disableInputAttributeSyncing" enabled
+      expect(textInputRef.current.value).toBe('');
+      expect(numberInputRef.current.value).toBe('');
+    } else {
```
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMForm-test.js]: Seems like this is undesired behavior? The flag isn't live anywhere and not being worked on so it's not blocking I believe.

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 1 above)
Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js:1005]: > I don't understand why it even exists - why is it expected that a controlled field can be overwritten in this way?

It seems to me it's just testing that you can input "0.0" which is a valid numeric input. Maybe I'm just not understanding what the new behavior is. Let's start by restoring the test and just updating the behavior.

These tests should include the actual user behavior we're simulating so that we can verify that the actual behavior is correct and it's just the test that needs updating. Can you do that for the new tests?

File: packages/react-dom/src/__tests__/ReactDOMForm-test.js (hunk 4 above)
Comment (vordgi) [reply packages/react-dom/src/__tests__/ReactDOMForm-test.js]: This condition was added after the `yarn test --variant=true` test failed. This flag is used in it...

Comment (vordgi) [reply packages/react-dom/src/__tests__/ReactDOMForm-test.js]: _And as I understand from other tests, it should work exactly like this_

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 1 above)
Comment (vordgi) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js:1005]: There's a very strange situation with the tests here. It turns out I removed the logic that set the defaultValue for number input on blur.

The check in this test is as follows - render with a controllable value of 0, focus, change the value to 0.0, blur.

After I removed the blur logic, this test failed. Which seems logical, but why does the logic of setting defaultValue play any role in this case? The field had a value of 0, how did changing defaultValue affect this value?

How should this work in real experience? That the user changed the value, then they left the field and it became 0.0?

Comment (vordgi) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js:1005]: It's very unfortunate that we can't conduct a normal test specifically in the browser env. I encountered this problem in submit tests and here it's similar. _Just out of curiosity, do you have any thoughts on setting up playwright or something like it here?_

File: packages/react-dom/src/__tests__/ReactDOMForm-test.js (hunk 4 above)
Comment (vordgi) [reply packages/react-dom/src/__tests__/ReactDOMForm-test.js]: And if this flag not being worked on - maybe we can remove it everywhere and forget about it forever?)

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 1 above)
Comment (vordgi) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js:1005]: Would it be sufficient to simply have a test where an input is rendered with a value of 0.0 and it's immediately checked that its value is indeed 0.0?

This sounds strange, but it's the only logical part in this test

File: packages/react-dom/src/__tests__/ReactDOMForm-test.js (hunk 4 above)
Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMForm-test.js]: > And if this flag not being worked on - maybe we can remove it everywhere and forget about it forever?)

That'd be a larger discussion. I wasn't implying this test should be removed. Just wanted clarification when we do resume work on this flag.

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 1 above)
Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js:1005]: > Just out of curiosity, do you have any thoughts on setting up playwright or something like it here?

Yeah but that's out of scope here. We can use Codesandboxes and Codesandbox Preview builds to verify the behavior.

It'd still be valueable for now to restore the test and change the assertions. We're currently shipping breaking changes so a subtle change to how DOM event listeners may observe events may be fine.

**Comment by vordgi** [comment]
_Just thoughts out loud, so as not to lose all these points and conclusions._

It turns out:

1. Inputs can be controlled or uncontrolled. Controlled ones are those that have a value. However, those that have a value but no change event will report an warning;
2. defaultValue is the base value of the input. It should be set as the value during rendering and will be restored when the field is reset (*if it's uncontrolled*);
3. If the field is controlled, it's expected that defaultValue will repeat the value for using the logic above;
4. Chrome has a bug where changing defaultValue can trigger validation (for number and email inputs), so defaultValue for number input is set on blur (*after change it will be on submit*);

Details:

- All this logic works on every change in the field;
- All this logic should work for number, text, email, date, tel and other text-like fields;
- For checkbox, everything should work the same, but with checked and defaultChecked props;
- For radio, the logic should track all fields with the current name in the form and control attributes considering the others;
- For the select element, it's the same as for radio, but with selected and defaultSelected props;
- For button, reset, image, file and other fields without input possibility - nothing should happen (*or is the value attribute removed?*)

What's strange

- In some situations, during input, the value attribute itself changes, in others it doesn't;
- Value changes work differently (*see above - why a text field can't change 0 to 0.0, but a number field can*);
- The logic still doesn't work on checkbox, radio, select even though there are separate conditions for these fields in the code and, for example, defaultChecked processing;
- The logic on defaultValue works on all field types, including checkbox, which supposedly doesn't need it;

All this logic is definitely needed on reset. But judging by the tests, it somehow participates in re-renders. But how? What kind of situation is it that I was in a field, entered a value into it, then a re-render occurred and the value was cleared. Apparently, the field was completely recreated, but can this happen for the same element in the same place?

Is all this really the right way? Maybe there's a possibility to leave native field handling? But the question is - where exactly is this logic needed. If in fact it's only needed during reset - we can look towards customizing the reset itself (resetting uncontrolled fields to the original defaultValue, leaving controlled fields unchanged).

File: packages/react-dom-bindings/src/client/ReactDOMInput.js [hunk 5]
```diff
@@ -418,26 +417,37 @@ export function restoreControlledInputState(element: Element, props: Object) {
...
   if (
-    // Focused number inputs synchronize on blur. See ChangeEventPlugin.js
-    type !== 'number' ||
-    getActiveElement(node.ownerDocument) !== node
+    type === 'button' ||
+    type === 'file' ||
+    type === 'hidden' ||
+    type === 'image' ||
+    type === 'reset' ||
+    type === 'submit'
```
Comment (sophiebits) [inline packages/react-dom-bindings/src/client/ReactDOMInput.js:433]: these are the types where setting .value sets the value attribute, yes?

should radio and checked be in this list as well? https://html.spec.whatwg.org/multipage/input.html#dom-input-value-default-on

File: packages/react-dom-bindings/src/client/ReactDOMInput.js [hunk 6]
```diff
@@ -418,26 +417,37 @@ export function restoreControlledInputState(element: Element, props: Object) {
...
-    }
+    return;
+  }
+
+  // In Chrome, assigning defaultValue to certain input types triggers input validation.
+  // For number inputs, the display value loses trailing decimal points. For email inputs,
+  // Chrome raises "The specified value <x> is not a valid email address".
+  // https://github.com/facebook/react/issues/7253
+  if (
+    (type === 'number' || type === 'email') &&
```
Comment (sophiebits) [inline packages/react-dom-bindings/src/client/ReactDOMInput.js:443]: does `url` have the same issue?

File: packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js [hunk 7]
```diff
@@ -92,6 +96,19 @@ function extractEvents(
   }
   const formInst = maybeTargetInst;
   const form: HTMLFormElement = (nativeEventTarget: any);
+
+  if (!disableInputAttributeSyncing) {
+    for (let i = 0; i < form.elements.length; i++) {
+      const input: HTMLInputElement = (form.elements.item(i): any);
+      if (
+        input.nodeName.toLowerCase() === 'input' &&
```
Comment (sophiebits) [inline packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js:104]:
```suggestion
        input.nodeName === 'INPUT' &&
```

**Review by sophiebits — COMMENTED** [review]
still need to think more about if this makes sense, but a couple questions

File: packages/react-dom-bindings/src/client/ReactDOMInput.js (hunk 6 above)
Comment (vordgi) [reply packages/react-dom-bindings/src/client/ReactDOMInput.js:443]: To be honest, I don't know. Even the number input doesn't have this problem now, so I based it on the previous comment (_I think in about a year, this difference for number input can be removed_)

File: packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js (hunk 7 above)
Comment (vordgi) [reply packages/react-dom-bindings/src/events/plugins/FormActionEventPlugin.js:104]: I just saw this particular format in another place and decided to repeat it (_but there's no point in this_)

https://github.com/facebook/react/blob/main/packages/react-dom-bindings/src/client/inputValueTracking.js

File: packages/react-dom-bindings/src/client/ReactDOMInput.js (hunk 5 above)
Comment (vordgi) [reply packages/react-dom-bindings/src/client/ReactDOMInput.js:433]: These are fields for which either dynamic value replacement is not expected, or there is no value at all.

For checkbox and radio, separate conditions with different processing are needed (_f.e. changing defaultChecked instead of defaultValue, and other specifics_). I mentioned above that if the current solution will be fine, I'll try to add processing for these types as well

Comment (vordgi) [reply packages/react-dom-bindings/src/client/ReactDOMInput.js]: Removed this condition so that only changes for email and number fields remain in PR

**Comment by vordgi** [comment]
@eps1lon @sophiebits

Is there anything else that needs to be done? I'm not closing the threads in case you have any more questions or a different perspective (_you know all the details better_)

So from the active question - should we bring back the logic with blur? At first glance, it seems unnecessary and redundant. But if you prefer more cautious changes, we can restore it and try to remove it in subsequent tasks.

_I think I've answered the rest of the questions, and I've also detailed all the issues and the current principle (might come in handy)_

**Comment by sophiebits** [comment]
Apologies this is still open — removing the stale label.
